README.md

```diff
@@ -0,0 +1,192 @@
+# Programming Arcana
+
+The Programming Arcana is an open source programming text book that focuses on teaching introductory programming concepts.
+
+The guiding principles are:
+
+- Transferable concepts are more important than language specifics
+- Build concepts on top of known concepts as much as possible
+- Focus on procedural programming initially, with a view to transitioning to objects
+- Build programs that make it easy to see what is happening in the code
+
+# Code formatter
+
+We use [_clang-format_](https://clang.llvm.org/docs/ClangFormat.html) as our code format for this project.
+
+## Installation
+
+### Linux
+
+_clang-format_ can be installed by:
+
+```bash
+sudo apt install clang-format
+```
+
+### macOS
+
+_clang-format_ can be installed by:
+
+```bash
+brew install clang-format
+```
+
+### Windows
+
+The executable file can be downloaded from [llvm org](https://llvm.org/builds/).
+
+### Other
+
+The _clang-format_ can also be installed with python and nodejs:
+
+```bash
+# python
+python3 -m pip install clang-format
+
+# nodejs
+npm install -g clang-format
+```
+
+The [VS code extension](https://marketplace.visualstudio.com/items?itemName=xaver.clang-format). (optional, clang-format required)
+
+## The format style
+
+The base format style used is [_google_](https://google.github.io/styleguide/cppguide.html).
+
+## Usage
+
+### Validate Code
+
+A python script from [Sarcasm](https://github.com/Sarcasm/run-clang-format) is used to validate the code.
+
+For example, you can use check if the file _./code/c/array/array-copy.c_ needs to be formatted.
+
+```bash
+# make the wrapped script executable
+chmod +x ./script/run-clang-format.py
+# use python instead of python3 if python version is 2
+./script/run-clang-format.py  ./code/c/array/array-copy.c
+```
+
+The output would be
+
+```
+--- ./code/c/array/array-copy.c (original)
++++ ./code/c/array/array-copy.c (reformatted)
+@@ -2,14 +2,13 @@
+
+ #include <string.h>
+
+-int main()
+-{
+-    int data[3] = {1, 2, 3};
+-    int other_data[3];
+-    int more_data[3];
+-
+-    memcpy(other_data, data, 3 * sizeof(int));
+-    memcpy(more_data, data, sizeof(data));
+-
+-    return 0;
++int main() {
++  int data[3] = {1, 2, 3};
++  int other_data[3];
++  int more_data[3];
++
++  memcpy(other_data, data, 3 * sizeof(int));
++  memcpy(more_data, data, sizeof(data));
++
++  return 0;
+ }%
+```
+
+\*Python3 is normally installed as default package for most operation systems. However, if you are using windows and do not want to install Python. You can use:
+`clang-format .\code\c\array\array-copy.c `, but it would not have the feature in the wrapped script e.g. color.
```

Comment:
```suggestion
\*Python3 is normally installed as a default package in most operating systems. However, if you are using Windows and do not want to install Python, you can use
`clang-format .\code\c\array\array-copy.c `. However,  it would not have the feature in the wrapper script, for example, coloured output.
```